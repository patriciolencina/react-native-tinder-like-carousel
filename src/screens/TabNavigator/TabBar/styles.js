import {StyleSheet} from 'react-native';
export default StyleSheet.create({
  containerSafe: {
    height: 44,
    backgroundColor: 'red',
  },
  container: {
    flexDirection: 'row',
    alignContent: 'center',
    height: 44,
    width: '100%',
    paddingHorizontal: 0,
    backgroundColor: 'red',
  },
  tabBarItem: {
    flex: 1,
    alignItems: 'center',
  },
  wrapIcon: {
    flex: 1,
    alignItems: 'center',
  },
  iconActive: {
    borderTopWidth: 3,
    borderColor: 'gray',
  },
  iconInactive: {
    borderTopWidth: 3,
    borderColor: 'blue',
  },
  textStyle: {
    marginTop: 20,
    color: 'gray',
    fontSize: 9,
    textAlign: 'center',
  },
  backgroundView: {
    position: 'absolute',
    flex: 1,
    height: '100%',
    width: 80,
    backgroundColor: 'gray',
  },
});
