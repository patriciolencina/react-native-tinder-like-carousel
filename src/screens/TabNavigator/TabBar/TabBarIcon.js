import * as React from 'react';
import {TouchableWithoutFeedback, View, Animated, Easing} from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';
import styles from './styles';
export default class TabBarIcon extends React.PureComponent {
  constructor(props) {
    super(props);
    this.animatedValue = new Animated.Value(0);
  }
  handleAnimation = () => {
    if (this.props.focused) {
      Animated.timing(this.animatedValue, {
        toValue: 1,
        duration: 200,
        easing: Easing.ease,
      }).start();
    } else {
      Animated.timing(this.animatedValue, {
        toValue: 0,
        duration: 50,
        easing: Easing.ease,
      }).start();
    }
  };
  onSelect = routeName => this.props.onPress(routeName);

  render() {
    this.handleAnimation();

    const {index, routeName} = this.props;
    let icon = '';

    switch (index) {
      case 0:
        icon = 'ios-home';
        break;

      case 1:
        icon = 'ios-heart';
        break;
    }
    const translateXObject = {
      translateX: this.animatedValue.interpolate({
        inputRange: [0, 1],
        outputRange: [0, index < 2 ? 2 : index === 2 ? 0 : -2],
      }),
    };
    return (
      <TouchableWithoutFeedback onPress={() => this.onSelect(routeName)}>
        <View style={[styles.wrapIcon]}>
          <Animated.View
            style={[
              styles.icon,
              {
                transform: [
                  {
                    translateY: this.animatedValue.interpolate({
                      inputRange: [0, 1],
                      outputRange: [0, -15],
                    }),
                  },
                  translateXObject,
                ],
              },
            ]}>
            <Icon name={icon} color="black" size={34} />
          </Animated.View>
          <Animated.Text
            style={[
              styles.textStyle,
              {
                transform: [
                  {
                    translateY: this.animatedValue.interpolate({
                      inputRange: [0, 1],
                      outputRange: [0, -40],
                    }),
                  },
                  translateXObject,
                ],
              },
            ]}>
            {routeName}
          </Animated.Text>
        </View>
      </TouchableWithoutFeedback>
    );
  }
}
