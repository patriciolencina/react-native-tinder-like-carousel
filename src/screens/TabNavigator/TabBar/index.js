import React, {useRef} from 'react';
import {SafeAreaView, View, Animated, Easing, Dimensions} from 'react-native';
import {path} from 'ramda';
import TabBarIcon from '../../../components/TabBarIcon';
import styles from './styles';
const screenWidth = Math.round(Dimensions.get('window').width);
const backgroundViewWidth = screenWidth / 2;

const TabBar = ({navigation = {}}) => {
  const animatedValue = useRef(new Animated.Value(0)).current;

  const routes = path(['state', 'routes'], navigation) || [];
  const navigationIndex = path(['state', 'index'], navigation);

  Animated.timing(animatedValue, {
    toValue: navigationIndex * backgroundViewWidth,
    duration: 200,
    easing: Easing.ease,
  }).start();

  return (
    <SafeAreaView style={styles.containerSafe}>
      <View style={styles.container}>
        <Animated.View
          style={[
            styles.backgroundView,
            {width: backgroundViewWidth},
            {
              transform: [
                {
                  translateX: animatedValue,
                },
              ],
            },
          ]}
        />

        {routes.map((route, index) => {
          return (
            <TabBarIcon
              key={route.key}
              routeName={route.routeName}
              onPress={currentRoute => navigation.navigate(currentRoute)}
              focused={navigationIndex === index}
              index={index}
            />
          );
        })}
      </View>
    </SafeAreaView>
  );
};

export default TabBar;
