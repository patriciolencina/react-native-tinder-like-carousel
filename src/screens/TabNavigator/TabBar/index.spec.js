import React from 'react';

import {shallow, ShallowWrapper} from 'enzyme';
import TabBar from './index';

const createTestProps = (props: Object) => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...props,
});

describe('TabBar', () => {
  describe('rendering', () => {
    let wrapper: ShallowWrapper;
    let props: any; // use type "any" to opt-out of type-checking
    beforeEach(() => {
      props = createTestProps({});
      wrapper = shallow(<TabBar {...props} />); // no compile-time error
    });

    it('should render', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
