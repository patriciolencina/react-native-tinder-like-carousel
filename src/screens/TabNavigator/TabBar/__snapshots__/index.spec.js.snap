// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabBar rendering should render 1`] = `
<SafeAreaView
  style={
    Object {
      "backgroundColor": "lightgray",
      "height": 44,
    }
  }
>
  <View
    style={
      Object {
        "alignContent": "center",
        "backgroundColor": "lightgray",
        "flexDirection": "row",
        "height": 44,
        "paddingHorizontal": 0,
        "width": "100%",
      }
    }
  >
    <AnimatedComponent
      style={
        Array [
          Object {
            "backgroundColor": "gray",
            "flex": 1,
            "height": "100%",
            "position": "absolute",
            "width": 80,
          },
          Object {
            "width": 375,
          },
          Object {
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          },
        ]
      }
    />
  </View>
</SafeAreaView>
`;
